File: Classes/Controls/DocumentView/Files/Cells/FilePreviewCollectionViewCell.m
//
//  FilePreviewCollectionViewCell.m
//  Dokie
//
//  Created by Ashamaz Shidov on 14/01/2020.
//  Copyright © 2020 Ashamaz Shidov. All rights reserved.
//

#import "FilePreviewCollectionViewCell.h"
#import "CoreDataInclude.h"
#import "CoreDataManager.h"
#import "FileImageView.h"
#import "UI.h"

#import <ReactiveObjC/ReactiveObjC.h>

@interface FilePreviewCollectionViewCell()

@property (nonatomic, strong) FileImageView *fileImage;

@property (nonatomic, strong, readonly) CoreDataManager *coreDataManager;

@end

@implementation FilePreviewCollectionViewCell

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self setup];
    }
    return self;
}

- (void)setup {
    self.fileImage = [FileImageView new];
    self.fileImage.translatesAutoresizingMaskIntoConstraints = NO;
    self.fileImage.contentMode = UIViewContentModeScaleAspectFill;
    self.fileImage.clipsToBounds = YES;
    [self.contentView addSubview:self.fileImage];
    
    UIView *shadowView = [UI shadowView];
    shadowView.alpha = 0;
    [self.contentView addSubview:shadowView];
    
    UIImageView *checkImage = [UI imageView];
    checkImage.image = [[UIImage imageNamed:@"checkmark.png"] imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
    [shadowView addSubview:checkImage];
    
    NSDictionary *metrics = @{@"imageSize"  :   @44};
    NSDictionary *views = NSDictionaryOfVariableBindings(_fileImage, shadowView, checkImage);
    [self.contentView addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|[_fileImage]|" options:0 metrics:nil views:views]];
    [self.contentView addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|[_fileImage]|" options:0 metrics:nil views:views]];
    [self.contentView addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|[shadowView]|" options:0 metrics:nil views:views]];
    [self.contentView addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|[shadowView]|" options:0 metrics:nil views:views]];
    [shadowView addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:[checkImage(imageSize)]-|" options:0 metrics:metrics views:views]];
    [shadowView addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:[checkImage(imageSize)]-|" options:0 metrics:metrics views:views]];
    
    @weakify(shadowView);
    [[[RACObserve(self, showCheckmark) distinctUntilChanged] deliverOnMainThread] subscribeNext:^(NSNumber *showCheckmark) {
        @strongify(shadowView);
        shadowView.alpha = showCheckmark.boolValue ? 1 : 0;
    }];
}

- (void)updateWithContent:(CDContent *)content {
    [self.fileImage updateWithContent:content];
}

#pragma mark - Get Set

- (CoreDataManager *)coreDataManager {
    return [CoreDataManager shared];
}

@end

File: Classes/Views/DocumentView/Files/Cells/FilePreviewCollectionViewCell.swift
//
//  FilePreviewCollectionViewCell.swift
//  Dokie
//
//  Created by Ashamaz Shidov on 24.05.2020.
//  Copyright © 2020 Ashamaz Shidov. All rights reserved.
//

import UIKit
import ReactiveSwift
import ReactiveCocoa

class FilePreviewCollectionViewCell: UICollectionViewCell {
    
    let showCheckmark = MutableProperty<Bool>(false)
    
    private let fileImage = UI.fileImageView()
    private let coreDataManager = CoreDataManager.shared

    override init(frame: CGRect) {
        super.init(frame: frame)
        setup()
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    private func setup() {
        fileImage.contentMode = .scaleAspectFill
        contentView.addSubview(fileImage)
        
        let shadowView = UI.shadowView()
        shadowView.alpha = 0
        contentView.addSubview(shadowView)
        
        let checkImage = UI.imageView()
        checkImage.image = UIImage(named: "checkmark.png")?.withRenderingMode(.alwaysTemplate)
        shadowView.addSubview(checkImage)
        
        let metrics = ["imageSize" : 44]
        let views = ["fileImage" : fileImage, "shadowView" : shadowView, "checkImage" : checkImage]
        contentView.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|[fileImage]|", options: NSLayoutConstraint.FormatOptions(rawValue: 0), metrics: metrics, views: views))
        contentView.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:|[fileImage]|", options: NSLayoutConstraint.FormatOptions(rawValue: 0), metrics: metrics, views: views))
        contentView.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|[shadowView]|", options: NSLayoutConstraint.FormatOptions(rawValue: 0), metrics: metrics, views: views))
        contentView.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:|[shadowView]|", options: NSLayoutConstraint.FormatOptions(rawValue: 0), metrics: metrics, views: views))
        shadowView.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:[checkImage(imageSize)]-|", options: NSLayoutConstraint.FormatOptions(rawValue: 0), metrics: metrics, views: views))
        shadowView.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:[checkImage(imageSize)]-|", options: NSLayoutConstraint.FormatOptions(rawValue: 0), metrics: metrics, views: views))
        
        showCheckmark
            .producer
            .take(during: reactive.lifetime)
            .observe(on: UIScheduler())
            .startWithValues { [unowned shadowView] show in
                shadowView.alpha = show ? 1 : 0
        }
    }
    
    func update(content: CDContent) {
        fileImage.update(content: content)
    }
    
}
